selfhost/typechecker.jakt

```diff
@@ -6061,148 +5991,296 @@ struct Typechecker {
+                | GenericEnumInstance(id) => Some(StructOrEnumId::Enum(id))
+            JaktString => Some(StructOrEnumId::Struct(.find_struct_in_prelude("String")))
+            GenericInstance(id, args) => match is_optional {
+                true => {
+                    let optional_struct_id = .find_struct_in_prelude("Optional")
+                    mut struct_id: StructOrEnumId? = None
+                    if not id.equals(optional_struct_id) {
+                        .error(format("Can't use ‘{}’ as an optional type in optional chained call", .get_struct(id).name), span)
+                    } else {
+                        found_optional = true
+                        struct_id = match .get_type(args[0]) {
+                            Struct(id)
+                                | GenericInstance(id) => Some(StructOrEnumId::Struct(id))
+                            Enum(id)
+                                | GenericEnumInstance(id) => Some(StructOrEnumId::Enum(id))
+                            else => {
+                                .error("Can't use non-struct type as an optional type in optional chained call", span)
+                                found_optional = false
+                                yield None
+                            }
+                        }
+                    }
+                    yield Some(struct_id ?? StructOrEnumId::Struct(optional_struct_id))
+                }
+                else => Some(StructOrEnumId::Struct(id))
             }
+            else => {
+                .error(format("no methods available on value (type: {})", .type_name(checked_expr_type_id)), checked_expr.span())
+                yield None
+            }
+        }
 
-            return match result! {
-                Return(x) => value_to_checked_expression(x, interpreter)
-                Throw(x) => {
-                    .error(
-                        format("Compiletime call failed: {}", x)
-                        span
-                    )
-                    yield checked_call
+        if is_optional and not found_optional {
+            .error(format("Optional chain mismatch: expected optional chain, found {}", .type_name(checked_expr_type_id)), checked_expr.span())
+        }
+
+        let function_id = match parent_id.has_value() {
+            true => .resolve_method_call(parent_id: parent_id!, call, span)
+            // skip any resolution since we already gave an error explaining why
+            // we cannot resolve a method here.
+            else => None
+        }
+
+        guard function_id.has_value() else {
+            return .checked_call_to_method_call(
+                expr: checked_expr
+                checked_call: .typecheck_unresolved_call(call, scope_id, safety_mode)
+                is_optional, span)
+        }
+
+        let callee = .get_function(function_id!)
+
+        let checked_call = .typecheck_resolved_call(
+            call, caller_scope_id: scope_id, span
```

Comment:
In cases where arguments are being split to multiple lines I think one argument per line is best.